olp-cpp-sdk-core/src/client/OlpClient.cpp

```diff
@@ -141,24 +170,27 @@ HttpResponse SendRequest(const http::NetworkRequest& request,
       [&]() {
         outcome = settings.network_request_handler->Send(
             request, response_body,
-            [&, interest_flag](http::NetworkResponse response) {
-              if (interest_flag->exchange(false)) {
+            [&, shared_state](http::NetworkResponse response) {
+              if (CheckState(shared_state)) {
                 network_response = std::move(response);
                 condition.Notify();
+                shared_state->store(State::Done);
               }
             },
-            [&headers, interest_flag](std::string key, std::string value) {
-              if (interest_flag->load()) {
+            [&headers, shared_state](std::string key, std::string value) {
+              if (CheckState(shared_state)) {
                 headers.emplace_back(std::move(key), std::move(value));
               }
             });
 
-        return CancellationToken([&, interest_flag]() {
-          if (interest_flag->exchange(false)) {
+        return CancellationToken([&, shared_state]() {
+          State state = State::Waiting;
+          if (shared_state->compare_exchange_weak(state, State::Cancelling)) {
```

Comment:
Be aware that this task can be cancelled anytime as it is attached to the CancellationContext given to the user. So it means we can also land here when state is `State::Processing` or `State::Done`.

Reply:
And we cannot cancel it in processing state or done. This is the same behavior as it was before.